# ВЫПУСКНОЙ ПРОЕКТ SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

## Содержание

[Оглавление](#intro)
1. [Описание проекта, цели и задачи](#intro1)
2. [Описание данных, общая информация](#intro2)
3. [Предоработка данных](#intro3)
4. [Исследовательский анализ данных](#intro4)
6. [Вывод](#intro5)

## [Описание проекта, цели и задачи](#intro1)

### Цель исследования — проанализировать приобретенную базу данных: информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 
Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Ход исследования

1. Опишем цели исследования.
2. Исследуем таблицы — выведем первые строки.
3. Сделайте по одному SQL запросу:
  * Посчитаем, сколько книг вышло после 1 января 2000 года;
  * Для каждой книги посчитаем количество обзоров и среднюю оценку;
  * Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключаем из анализа брошюры;
  * Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками;
  * Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
4. Выведем результат каждого запроса.
5. Опишем выводы по каждому запросу. 

## [Описание данных, общая информация](#intro2)

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<a href="https://ibb.co/mXt1LpS"><img src="https://i.ibb.co/Jy20fZC/Untitled-33.png" alt="Untitled-33" border="0"></a>

### Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

## [Предоработка данных](#intro3)

In [2]:
# Выведем данные о книгах из таблицы `books`

query = '''
SELECT *
FROM books
'''

In [3]:
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
# Выведем данные об авторах из таблицы `authors`

query = '''
SELECT *
FROM authors
'''

In [5]:
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [6]:
# Выведем данные об издательствах из таблицы `publishers`

query = '''
SELECT *
FROM publishers
'''

In [7]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [8]:
# Выведем данные о пользовательских оценках книг `ratings`

query = '''
SELECT *
FROM ratings
'''

In [9]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [10]:
# Выведем данные о пользовательских обзорах на книги `reviews`

query = '''
SELECT *
FROM reviews
'''

In [11]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## [Исследовательский анализ данных](#intro4)

**Посчитаем, сколько книг вышло после 1 января 2000 года**

In [12]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '01.01.2000'
'''

In [13]:
pd.io.sql.read_sql(query, con = engine)

count
0    819

Посе 1 января 2000 года вышло 819 книг.

**Для каждой книги посчитаем количество обзоров и среднюю оценку**

In [14]:
query = '''
SELECT b.title
      ,reviews_query.count_reviews AS count_reviews
      ,AVG(rat.rating) AS avg_rating
      
FROM books AS b

LEFT JOIN (SELECT book_id
                  ,COUNT(review_id) AS count_reviews
           FROM reviews
           GROUP BY 1
          ) AS reviews_query ON reviews_query.book_id = b.book_id
          
LEFT JOIN ratings AS rat ON rat.book_id = b.book_id

WHERE count_reviews IS NOT NULL
GROUP BY 1, 2
ORDER BY 2 DESC
'''

In [15]:
pd.io.sql.read_sql(query, con = engine)

title  count_reviews  \
0                              Twilight (Twilight  #1)              7   
1                                      Eat  Pray  Love              6   
2    Harry Potter and the Chamber of Secrets (Harry...              6   
3    Harry Potter and the Prisoner of Azkaban (Harr...              6   
4                            Outlander (Outlander  #1)              6   
..                                                 ...            ...   
989                                The Social Contract              1   
990  To Green Angel Tower (Memory  Sorrow  and Thor...              1   
991               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1              1   
992                              Wicked: The Grimmerie              1   
993                  Winter Prey (Lucas Davenport  #5)              1   

     avg_rating  
0      3.662500  
1      3.395833  
2      4.287500  
3      4.414634  
4      4.125000  
..          ...  
989    3.500000  
990    4.500000  
991    2.500000  
992    3.500000  
993    4.500000  

[994 rows x 3 columns]

Из 1000 книг 994 книги имеют обзоры. Самое большое колиество 7 обзоров написано на книгу Twilight (Twilight #1), средний рейтинг у неё 3,66. 6 книг не имеют обзоров.

**Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключаем из анализа брошюры**

In [16]:
query = '''
SELECT p.publisher
       ,COUNT(b.book_id) AS count_books
FROM books AS b
LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
'''

In [17]:
pd.io.sql.read_sql(query, con = engine)

publisher  count_books
0  Penguin Books           42

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги.

**Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками**

In [18]:
query = '''
SELECT a.author
       ,AVG(ratings_query.avg_rating) AS avg_rating
FROM books AS b
LEFT JOIN (
           SELECT book_id
                  ,AVG(rating) AS avg_rating
                  ,COUNT(rating_id) AS count_rating
            FROM ratings
            GROUP BY 1
          ) AS ratings_query ON ratings_query.book_id = b.book_id
LEFT JOIN authors AS a ON a.author_id = b.author_id
WHERE count_rating >= 50
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
'''

In [19]:
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré - Джоан К. Роулинг/Мэри Гранпре. Средняя оценка ее книг составляет 4.28 балла.

**Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.**

In [20]:
query = '''
WITH
ratings_query AS (
                     SELECT r.username
                            ,COUNT(rating_id) AS count_ratings
                     FROM ratings AS r
                     GROUP BY 1
                     HAVING COUNT(rating_id) > 48
                     ),
reviews_query AS (
                     SELECT rev.username
                            ,COUNT(review_id) AS count_reviews
                     FROM reviews AS rev
                     GROUP BY 1
                     )
SELECT AVG(count_reviews) AS avg_reviews_count
FROM ratings_query
LEFT JOIN reviews_query ON ratings_query.username = reviews_query.username
'''

In [21]:
pd.io.sql.read_sql(query, con = engine)

avg_reviews_count
0               24.0

Пользователи, которые поставили больше 48 оценок очень активно пишут и отзывы. В среднем ими написано 24 обзора! Настоящие книголюбы)

## [Вывод](#intro5)

Приобретенная база данных содержит 1000 наименований книг, 636 авторов, 340 издательств. Пользователи проставили 6456 оценок, и написали 2793 обзора.

Посе 1 января 2000 года вышло 819 книг.

Из 1000 книг 994 книги имеют обзоры. Самое большое колиество 7 обзоров написано на книгу Twilight (Twilight #1), средний рейтинг у неё 3,66. 6 книг не имеют обзоров.

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги.

Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré - Джоан К. Роулинг/Мэри Гранпре. Средняя оценка ее книг составляет 4.28 балла.

Пользователи, которые поставили больше 48 оценок очень активно пишут и отзывы. В среднем ими написано 24 обзора! Настоящие книголюбы)